### DISCUSSION TITLE
Cliff Aster does not see attachments from iOS backup?

### DISCUSSION THREAD

**Discussion by 2600box** [ReagentX/imessage-exporter#117] [General]
Hello, when I run `./imessage-exporter-x86_64-apple-darwin -a iOS -p /Volumes/BACKUP/00008120-001644390CEB401E -d`

It shows that it cannot find any attachments? Is there something more that I need to do?

```
Building cache...
[1/4] Caching chats...
[2/4] Caching chatrooms...
[3/4] Caching participants...
[4/4] Caching reactions...
Cache built!

iMessage Database Diagnostics

Contacts with more than one ID: 25
Messages not associated with a chat: 23
Missing attachment data:
    Missing files: 16817
Duplicated contacts: 127
Duplicated chats: 2
Done!
```

**Comment by ReagentX** [comment]
It definitely should, and in my local getting against an unencrypted backup of an iOS 16.4.1 device, it does. Can you provide more details on the backup you are trying to read from?

**Comment by ReagentX** [comment]
This is actually a bug in the diagnostic code. If you create an export the attachments should show up. I am tracking this in #118.

**Comment by 2600box** [comment]
Ah, that it is good to know. I should have tried a proper run. That said, I still have a lot of missing attachments that I can't figure out. 

Can you tell me where does `imessage-exporter` get the strings for the backup files? Because I can't find them in Manifest.db or chat.db

`for X in $(sqlite3 Manifest.db .tables) ; do sqlite3 Manifest.db "SELECT * FROM $X;" | grep -C 5 >/dev/null '9524a9b201ef417e616fd4577705bbff55f30334' && echo $X; done`

`for X in $(sqlite3 ./3d/3d0d7e5fb2ce288813306e4d4636395e047a3d28 .tables) ; do sqlite3 ./3d/3d0d7e5fb2ce288813306e4d4636395e047a3d28 "SELECT * FROM $X;" | grep -C 5 >/dev/null '9524a9b201ef417e616fd4577705bbff55f30334' && echo $X; done`

**Comment by ReagentX** [comment]
The path to an attachment is stored in the `filename` column in the `attachments` table. For iOS, we use the following algorithm to find the file in the backup:

https://github.com/ReagentX/imessage-exporter/blob/30e9c17e63584d17a157ccaaf606dc3f458c4b0c/imessage-database/src/tables/attachment.rs#L211-L238

Once that is resolved, the exporter uses the following logic to copy the file (if enabled):

https://github.com/ReagentX/imessage-exporter/blob/30e9c17e63584d17a157ccaaf606dc3f458c4b0c/imessage-exporter/src/exporters/html.rs#L416-L488

**Reply by 2600box** [reply]
Hmmm, so when I have lines like `Attachment not found at specified path: "/Volumes/BACKUP/00008120-001644390CEB401E/95/9524a9b201ef417e616fd4577705bbff55f30334"` how can I check which file this should be referring to?

**Reply by ReagentX** [reply]
It means that your backup is located at `/Volumes/BACKUP/00008120-001644390CEB401E` and inside should be a file at `95/9524a9b201ef417e616fd4577705bbff55f30334`. If there isn't, my first guess is that it is not stored on the phone (likely not synced from iCloud). If the file isn't at the specified location there isn't really anything we can do.

**Reply by 2600box** [reply]
That part I understand, but I would like to track these files down. So, a more verbose message specifying the expected attachment from the db, for example `~/Library/SMS/Attachments/bf/15/ZZZ/IMG_9162.HEIC`

In this example I have no idea what file `9524a9b201ef417e616fd4577705bbff55f30334` is referring to and I could not figure it out...

**Reply by ReagentX** [reply]
`9524a9b201ef417e616fd4577705bbff55f30334` is what the filename was hashed to in the backup. The original filename is still accessible via the `filename()` method on the `Attachment`, but in the context of an iOS backup, that path never actually exists.

When parsing iOS backups, I can add that original filepath to the `Attachment not found at specified path` log, but again, since it doesn't actually exist, I think that is of dubious value.

**Reply by 2600box** [reply]
I understand if it is not a priority, but to explain why it would be useful for me. To have the information on missing files, specifically filename, path and total_bytes I could clearly identify the gaps and work on filling those gaps from other backups. 

I am sure it would be appreciated by anyone finding they are missing attachments. Thanks!

**Reply by ReagentX** [reply]
That is really good feedback, I will improve the error messages here in the next release.
